[Auditbeat] Login metricset

cwurm opened the pull request:
This adds the `login` metricset to the [Auditbeat system module](https://github.com/elastic/beats/issues/8725). It's the last of the six initial metricsets. It only works on Linux, and detects not just user logins and logouts, but also system boots and shutdowns.

It works by reading the `/var/log/wtmp` and `/var/log/btmp` file (and rotated files) present on Linux systems. In reading a file, it is similar to Filebeat, except that UTMP is a binary format, so reading happens using a binary Go reader. See [utmp(5)](http://man7.org/linux/man-pages/man5/utmp.5.html) for the format of that file.

The logic is roughly as follows:

1. The config parameters `login.utmp_file_pattern` and `login.btmp_file_pattern` will contain the pattern matching the wtmp (good logins, as well as system shutdowns and boots) and btmp (bad/failed logins) files and rotated files (if desired). The defaults are `/var/log/wtmp*` and `/var/log/btmp*`. These are expanded using `filepath.Glob` and the files are sorted lexicographically in reverse order (i.e. `/var/log/wtmp.1` will come before `/var/log/wtmp`) so that we read older login records first - reading in order is required for matching login and logout records, see next steps.
2. Every `Fetch` it checks for new entries: Any new files are read from the beginning, while known files are read from a saved offset. To that purpose, the last offset per file is saved and persisted to disk in `beat.db`. A new file is one that has an unknown inode, but files are also read completely if their `newSize < oldSize` for some reason (that should make it work with any potential inode reuse - very unlikely since this will never read a lot of files but still possible).
3. New UTMP records are converted to events (`LoginRecord` in the code). Boot and shutdown events are fairly straightforward, user login and logout events have to be matched using their `tty`- so there is a `loginSessions` map that stores logins to enrich the logouts, and is also persisted to disk.

Note: This dataset also introduces `event.origin` containing the file the event came from, e.g. `/var/log/wtmp.1`. In other cases, it would be something like `procfs` or `netlink`. It's useful to know where information comes from, e.g. to know how reliable it is.

cwurm commented on x-pack/auditbeat/module/system/login/config.go:
```diff
@@ -0,0 +1,19 @@
+	UtmpFilePattern string `config:"login.utmp_file_pattern"`
+}
+
+// Validate validates the host metricset config.
```
👍

cwurm commented on x-pack/auditbeat/module/system/login/config.go:
```diff
@@ -0,0 +1,19 @@
+
+// Config defines the metricset's configuration options.
+type Config struct {
+	UtmpFilePattern string `config:"login.utmp_file_pattern"`
```
Thanks! done

cwurm replied:
Ah, that actually forces the user to specify it. It has a default, so it should be fine for the user not to.

tsg commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	if loginRecord.Hostname != "" && loginRecord.Hostname != loginRecord.IP.String() {
+		event.RootFields.Put("source.hostname", loginRecord.Hostname)
```
To be sure, this is the hostname from where the user logs in remotely, right? It's not the host on which the user logs in.

I wonder if `source.hostname` is the correct ECS field for this. It doesn't show up explicitly in ECS (only `source.ip`), but it does seem to make sense to me.

cwurm replied:
Correct, it's the remote hostname. Since we already have `source.ip` I thought it makes sense. There's even `[source|destination].domain` in ECS already. 

I think other data sources could use this as well, e.g. flows where there might be a `source.hostname` and a `destination.hostname`.

cwurm replied:
@andrewkroh Because the actual hostname might not be what another host sees? E.g. `webserver001` might be the hostname, but to the outside world it appears as `web.elastic.co`?

tsg commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+			loginRecord.Hostname, loginRecord.IP)
+	}
+
+	event.RootFields.Put("event.summary", eventSummary)
```
I think we're going with `message` instead of `event.summary` for this. @webmat @MikePaquette, correct? Or do you prefer to do the rename in a follow up PR for the whole module?

cwurm replied:
Since this is the first time we're introducing this field we can settle on a name in this PR. I personally would want to keep it as `summary`, because it's different from `message` and there is a place for both of them in the same document. In the PR description above I'm arguing the details (pasting here):

event.summary contains a human-readable summary of the event, e.g. Login by user vagrant (UID: 1000) on pts/0 (PID: 9355) from 10.0.2.2 (IP: 10.0.2.2).. It doesn't add any information that is not already in any of the other fields, it's there to make it easy to consume this data together with data from the other system module metricsets e.g. in a table in Kibana. I chose not to use the message base field from ECS, since in my mind a message is something "real" that comes from outside the Beat e.g. a log message. summary is not original/authoritative and doesn't contain any new information. It can be removed without losing anything. It could be used to summarize a message field as well, e.g. the unwieldy access log messages from web servers.

tsg replied:
The advantage of `message` is that it would work OOB in the log viewer and similar UIs. We could maybe change the log viewer to also look for this field. Something to discuss.

cwurm replied:
As discussed, I've changed to using `message` instead of `event.summary` since that's the ECS standard.

tsg commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
```
Mentioning it mostly for awareness: if I'm not wrong, the Auditbeat `report.Event()`doesn't guarantee sending at least once. If Elasticsearch is down, for example, it will retry 3 times, and then drop the event. This means that we can potentially lose login events, despite saving the state. I think we could solve this after the Filebeat refactoring, but I suggest documenting this as a limitation for now.

cwurm replied:
Hm, that's a shame. It guess this affects all metricsets, and we might lose process starts, socket connections, password changes, audit events, FIM events, etc.

The [internal queue](https://www.elastic.co/guide/en/beats/auditbeat/master/configuring-internal-queue.html) does not keep events until they can be sent (or it overflows)? That would seem desirable.

tsg replied:
It does, and one can also use spooling to disk to make the queue larger. But Filebeat offers an extra guarantee by waiting for the ACK and only afterwards update the state on disk. This is not currently possible in Auditbeat. That’s the difference I wanted to highlight.

tsg commented on x-pack/auditbeat/module/system/login/login_test.go:
```diff
@@ -0,0 +1,36 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package login
```
This one also needs a build tag for linux only, otherwise tests fail on macOS.

cwurm replied:
👍  - done

tsg commented:
We should think through what the experience is on non-Linux systems. I'm thinking two things:

* If you enable the `login` module on macOS/Windows, you should get an unimplemented error in the logs
* The default windows/macOS configs should not enable this metricset.

Is the first point already happening? I couldn't test because of the `user.go` error.

cwurm commented:
>     * If you enable the `login` module on macOS/Windows, you should get an unimplemented error in the logs

That should be what's happening (in [login_other.go](https://github.com/elastic/beats/blob/91b46b0c4562dc5cc9bd19406be80dd0660d7adc/x-pack/auditbeat/module/system/login/login_other.go#L28)). I've added some extra `cgo` build flags in case it's not enabled when building.

>     * The default windows/macOS configs should not enable this metricset.

It shouldn't - there's an if/else around it [in the `config.yml.tmpl`](https://github.com/elastic/beats/blob/91b46b0c4562dc5cc9bd19406be80dd0660d7adc/x-pack/auditbeat/module/system/_meta/config.yml.tmpl#L8).

> I couldn't test because of the `user.go` error.

I opened https://github.com/elastic/beats/pull/9368 to address that - it compiles for me after applying that.

I've changed to using `message` as well.

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+	if utAddrV6[1] != 0 || utAddrV6[2] != 0 || utAddrV6[3] != 0 {
+		// IPv6
+		b := make([]byte, 16)
+		binary.LittleEndian.PutUint32(b[:4], utAddrV6[0])
```
What does this mean for big-endian architectures? Will this code work?

cwurm replied:
Hm, good point and I don't know. Do we support any big-endian systems?

adriansr replied:
I'm surprised Go doesn't have a binary.MachineEndian.

adriansr replied:
```
package main

import (
	"encoding/binary"
	"fmt"
	"unsafe"
)

var MachineEndian = getByteOrder()

func getByteOrder() binary.ByteOrder {
	var b [2]byte
	*((*uint16)(unsafe.Pointer(&b[0]))) = 1
	if b[0] == 1 {
		return binary.LittleEndian
	}
	return binary.BigEndian
}

func main() {
	var b [4]byte
	MachineEndian.PutUint32(b[:], 0x12345678)
	fmt.Printf("%x %x %x %x\n", b[0], b[1], b[2], b[3])
}

```

cwurm replied:
Just revisiting this. I'm thinking that it might not be likely that this would be used on a big-endian system that uses UTMP login records. In the interest of keeping the code simple I'd leave it as is. If there's a need to change things we always can.

@adriansr `getByteOrder()` looks like a useful function that Go itself should have, indeed. Or maybe we should have it in a shared location within Beats so code like this here can use it easily.

andrewkroh replied:
I think it's likely that someone will use it on big-endian. We've had several contributions to go-libaudit to make auditbeat work on big-endian.

cwurm replied:
Ok, I've added support for it. Thanks @adriansr for the code snippet!

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+		inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+
+		fileRecord, isKnownFile := r.fileRecords[inode]
+		var oldSize int64
```
You can skip the isKnownFile check here. The value of Size will be 0 if it's not found in the map.
```suggestion
		oldSize := fileRecord.Size
```

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading file %v", path)
+			} else if len(utmpRecords) == 0 {
+				return nil, fmt.Errorf("unexpectedly, there are no new records in file %v", path)
```
There's an `errors.Errorf` function that behaves similar to `errors.Wrap` in that it records stack information about where the error occurred. Also it might be useful for debugging purposes to include the number of bytes that were read (or the old and new sizes).

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+// ReadAfter reads a UTMP formatted file (usually /var/log/wtmp*)
+// and returns the records after the provided last known record.
+// If record is nil, it returns all records in the file.
+func (r *UtmpFileReader) readAfter(path string, lastKnownRecord *Utmp) ([]Utmp, error) {
```
The format of this utmp struct seems pretty simple. I think you could use a pure Go decoder and then not have to deal with reading the full file each time. Currently there's no way to seek to the last offset so you have read the whole file over.

For example: https://github.com/neko-neko/utmpdump/blob/master/utmp/utmp.go

cwurm replied:
Changed to pure Go decoder.

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+// but will return -1 instead.
+func lookupUsername(username string) int {
+	if username != "" {
+		user, err := user.Lookup(username)
```
This is probably fine for low volumes. But we have seen from the auditd module that this call is expensive. Some caching could be a useful future enhancement.

urso commented on x-pack/auditbeat/module/system/login/config.go:
```diff
@@ -0,0 +1,19 @@
+	return nil
+}
+
+var defaultConfig = Config{
```
Note: As configs tend to get more complex newer code should better use functions to prepare an initial config object. This reduces the risk of accidentily sharing configs if references (e.g. map or pointers) are used in the defaults.

This is no issue here, but some recommendation to keepin in mind in general.

urso commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+	eventTypeEvent = "event"
+
+	// LoginRecordTypeBoot represents a system boot.
+	LoginRecordTypeBoot = "boot"
```
+1.

- do not export the symbol if not required
- in go consider the package name to be part of the symbols full name. `login.LoginRecordTypeX` could become `xRecord` or `xRecordType`.

urso commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+	IP        *net.IP
+	Timestamp time.Time
+	Origin    string
+}
```
Same, long name and a many exported symbols that we might not need somewhere else.

cwurm replied:
`LoginRecord` is encoded and decoded by `gob` so it needs to be exported, unfortunately.

urso commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,193 @@
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
```
Calling save to disk here can lead to data loss between restarts. In Winlogbeat/Filebeat we fill out the `Private` in the event and write the state update after an event has been ACKed by the output. The new registry file format will be more similar to a small in-memory key value store allowing you to reuse the pattern used in filebeat/winlogbeat.

Another potential issue might be the bucket name. The bucket name and therefore the actual storage for state is unique for this module, but not per instance. One must not have more then 2 instances collecting a file, or you will have 2 go-routines updating global state at the same time.

cwurm replied:
Under which circumstances would the data loss happen? Even in case of a graceful shutdown? I suspect this can affect all metricsets (even the `auditd` module?), so at some point we might want to come up with a solution for all of Auditbeat (but maybe not now?).

Two Auditbeat instances using the same bucket concurrently would be a problem for sure. I would think that we should not allow two instances to use the same bucket, or even the same data path.

tsg replied:
I wouldn't block this metricset on the potential data loss. We can do better in the future, but I agree with @cwurm that from the user perspective it's the same guarantees as the other Auditbeat modules, so they won't expect more.

urso replied:
> Under which circumstances would the data loss happen? Even in case of a graceful shutdown?

Yes.

> I wouldn't block this metricset on the potential data loss.

OK

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+
+// FileRecord represents a UTMP file at a point in time.
+type FileRecord struct {
+	Inode    Inode
```
In filebeat we also store the device id.

cwurm replied:
I think for now it's unlikely that UTMP files would be stored on separate file systems. I.e. it would be mean different parts of `/var/log` are on different devices. I'm not even sure that's possible.

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+	}
+
+	// Sort paths in reverse order (oldest/most-rotated file first)
+	sort.Sort(sort.Reverse(sort.StringSlice(paths)))
```
The purpose of the loops seems to be filtering. e.g.:

```
filtered := paths[:0]
for _, path := range paths {
  if ok {
    filtered = append(filtered, path)
  }
}

return filtered
```

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+	reachedNewRecords := (lastKnownRecord == nil)
+	var utmpRecords []Utmp
+	for {
+		utmpC, err := C.getutent()
```
Can there be multiple instances of this module or the reader?

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+	var utmpRecords []Utmp
+	for {
+		utmpC, err := C.getutent()
+		if err != nil {
```
what if err == io.EOF?

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+}
+
+// ReadNew returns any new UTMP entries in any files matching the configured pattern.
+func (r *UtmpFileReader) ReadNew() ([]LoginRecord, error) {
```
How big do we expect these files to be? Looks like this function will load all available logs into memory on first run.

cwurm replied:
So far I've assumed they're usually quite small, but that might not be true in all cases. The `/etc/logrotate.conf` on my Ubuntu VM rotates it monthly. On that VM that's getting a number of daily logins/logouts, they're <100K. On a real-world machine with a lot of logins, I assume it could be more, I wonder how many records it would take to be significant.

Nevertheless, I guess we could use a channel between `Fetch` and `ReadNew` and output records as events as they come.

urso replied:
This is more or less what filebeat does as well. Have one go-routine per file and forward event by event. That is, the module/input is more or less active all the time. Not sure how well this fits the 'Fetch' semantics in the metricbeat framework. But a bounded channel can indeed help keeping memory usage in place. Given the complexity this introduces I wonder if that's really a problem.

cwurm replied:
I've changed to using channels now, and storing offsets of the files (instead of re-reading until the last read record). So hopefully that will make it scalable even if the files are very large.

urso commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+				r.log.Warnf("Unexpectedly, the file %v did not contain the saved login record %v - reading whole file.",
+					path, *lastKnownRecord)
+
+				return r.readAfter(path, nil)
```
Looks like we are ignoring all messages read so far if we end up here (e.g. we read while copytruncate like logrotation happens). We still want to publish known records? Alternatively we can just error and return here, but rely on fetch to reset the file offset and read new contents the next time Fetch is executed.

tsg replied:
But can `wtmp` be copy-truncated? I thought it's guaranteed to be renamed with moves. I wouldn't want to bring all the complexity of Filebeat's corner cases in here :)

urso replied:
Rename with move should be ok. But them we don't get into this code path.

Instead of having a tail recursive call we could use a for loop and reset some state and use `continue` in order to jump to beginning of the reader and not drop events. But then it's an edge case that normally should not occur.

cwurm replied:
The code is now using offsets and if there is some problem in jumping to the stored offset it will jump back to the beginning and re-read the file. This is maybe a bit overzealous, but I think at least for now I'd prefer it if events were sent twice rather than not at all. If there are any problems we can always change.

cwurm commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+			} else {
+				return nil, errors.Wrapf(err, "unexpected error when looking up file %v", path)
+			}
+		} else if fileInfo.Sys() == nil {
```
You're right - `fileInfo.sys` is filled directly with `syscall.Stat(name, &fs.sys)`.

cwurm commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+		for {
+			fileRecord := new(FileRecord)
+			err = decoder.Decode(fileRecord)
+			if err == nil {
```
In general, yes. Here, I couldn't use a `break` anymore to get out of the surrounding for loop and would have to introduce a label, or a stop variable. I think that would make it less readable.

cwurm commented on x-pack/auditbeat/module/system/login/utmp.go:
```diff
@@ -0,0 +1,571 @@
+// deleteOldRecords cleans up old file records where the inode no longer exists.
+func (r *UtmpFileReader) deleteOldRecords(fileInfos map[string]os.FileInfo) {
+	for savedInode := range r.fileRecords {
+		found := false
```
I think that would delete all `fileRecords`? The `break` only breaks the inner loop, so the `delete` runs every time.

cwurm commented:
I've rebased the code on `master` and made a number of changes. Should be ready for another review. I hopefully addressed all the comments, but better take a look.

High-level changes:

1. Now also reads failed logins from `/var/log/btmp` and fills `event.outcome` accordingly.
2. Uses a pure Go decoder instead of C library functions.
3. Stores and jumps to offsets instead of re-reading files to the last read entry.
4. Uses event and error channels to keep memory consumption constant and small no matter how large the files are.

I'm afraid the changes I've made are pretty extensive and there's probably no point in looking at just those. Sorry it's so long again.

Follow-up:
- [ ] Add a Kibana dashboard
- [ ] Propose adding `event.origin` and `user.terminal` to ECS
- [ ] Better tests, esp. some with sample `wtmp` and `btmp` files

webmat commented on x-pack/auditbeat/module/system/_meta/fields.yml:
```diff
@@ -4,7 +4,25 @@
+  - name: event
+    type: group
+    fields:
+    - name: origin
```
@ruflin I think this would make a great addition to ECS for later. Maps well to the semantics of Filebeat 6.x `source`.

ruflin replied:
Let's open issues in ECS and best with some value examples 👍

webmat commented on x-pack/auditbeat/module/system/_meta/fields.yml:
```diff
@@ -4,7 +4,25 @@
+  - name: user
+    type: group
+    fields:
+    - name: terminal
```
@ruflin Same here. I've also started using this in Filebeat's auditd module. I also think this is something we could consider in ECS.

webmat commented on x-pack/auditbeat/module/system/login/_meta/data.json:
```diff
@@ -0,0 +1,36 @@
+    "system": {
+        "audit": {
+            "login": {
+                "tty": "pts/1"
```
`user.terminal`

andrewkroh commented:
Can you please update the PR description to reflect the latest state of the PR (like how it doesn't use C).

andrewkroh commented on auditbeat/docs/fields.asciidoc:
```diff
@@ -6070,6 +6070,28 @@ These are the fields generated by the system module.
+--
+
+
+*`user.terminal`*::
```
I think it would be good to open a PR to elastic/ecs suggesting that this be added. I think the auditd module gives tty info.

cwurm replied:
It does. https://github.com/elastic/beats/pull/10192 will fill `process.terminal` for the Filebeat auditd module. I wonder if we should use the same field, and whether it should be `user.terminal` or `process.terminal`. @andrewkroh @webmat what do you think?

webmat replied:
Yeah in my PR I've been mapping both `auditd.log.tty` and `auditd.log.terminal` to `process.terminal`, thinking it was inconsistency on the auditd side. This makes me wonder if one is actually user-specific and one process-specific. I'll dig into this some more for my PR.

But I do think `user.terminal` makes sense, if you have terminal info for users.

webmat replied:
Found some light reading for the week-end: [Auditd field dictionary](https://github.com/linux-audit/audit-documentation/blob/master/specs/fields/field-dictionary.csv)

andrewkroh commented on x-pack/auditbeat/module/system/login/config.go:
```diff
@@ -0,0 +1,20 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux,cgo
```
This file isn't linux or cgo specific.

andrewkroh commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,234 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux,cgo
```
Does not need to be limited to linux,cgo? From what I understand the same file format is used on a few systems (like maybe some bsd flavors or solaris).

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp_c.go:
```diff
@@ -0,0 +1,105 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux,cgo
```
Being pure Go this is no longer needs to be tagged with linux,cgo.

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp_c.go:
```diff
@@ -0,0 +1,105 @@
+func ReadNextUtmp(r io.Reader) (*Utmp, error) {
+	utmpC := new(utmpC)
+
+	err := binary.Read(r, binary.LittleEndian, utmpC)
```
Is the endian part of the file spec or is this system specific?

andrewkroh commented on x-pack/auditbeat/module/system/login/login.go:
```diff
@@ -0,0 +1,234 @@
+func loginMessage(loginRecord *LoginRecord) string {
+	var message string
+
+	userString := fmt.Sprintf("by user %v (UID: %d) on %v (PID: %d) from %v (IP: %v)",
```
From an allocation perspective it might be better to use a `strings.Builder` + `fmt.Fprintf` or refactor so that there's only one `fmt.Sprintf`. You'll probably have to swap the order around a bit. For `bootRecord` and `shutdownRecord` you can do a straight `return "System ..."`.

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp_c.go:
```diff
@@ -0,0 +1,105 @@
+
+// Possible values for the ut_type field. See utmp(5).
+const (
+	EMPTY         = 0
```
You could create a `type UtType int16` for these and use that in your structs.

andrewkroh commented on x-pack/auditbeat/tests/system/test_metricsets.py:
```diff
@@ -21,6 +21,23 @@ def test_metricset_host(self):
         self.check_metricset("system", "host", COMMON_FIELDS + fields, warnings_allowed=True)
 
+    @unittest.skipUnless(sys.platform == "linux2", "Only implemented for Linux")
+    def test_metricset_login(self):
```
This test should be restricted to little endian architectures (or some other constraint) because `tests/files/wtmp` is from a LE machine.

cwurm replied:
done

andrewkroh commented on x-pack/auditbeat/module/system/login/utmp_c.go:
```diff
@@ -0,0 +1,120 @@
+// Possible values for UtType.
+const (
+	EMPTY         UtType = 0
+	RUN_LVL              = 1
```
I think you need to specify `UtType` for each of these. https://play.golang.org/p/Y169ZzumpVT

cwurm replied:
done

andrewkroh commented on x-pack/auditbeat/module/system/login/login_test.go:
```diff
@@ -0,0 +1,40 @@
+	return map[string]interface{}{
+		"module":                  "system",
+		"datasets":                []string{"login"},
+		"login.wtmp_file_pattern": "../../../tests/files/wtmp",
```
Due to endianness this test case should be restricted to the architecture of the host that this test data was sourced from.

cwurm replied:
done

adriansr commented on x-pack/auditbeat/module/system/login/utmp_c.go:
```diff
@@ -0,0 +1,120 @@
+
+var byteOrder = getByteOrder()
+
+func getByteOrder() binary.ByteOrder {
```
Turns out gosigar has a GetEndian() method for this.

https://github.com/elastic/beats/blob/f28cf30de74572b9323d616f16fffed177ba99c1/vendor/github.com/elastic/gosigar/sys/endian.go#L8

cwurm replied:
Oh nice, thanks!
